Add a chapter on Compute Shaders (#366)
KhronosGroup/Vulkan-Guide: chapters/compute_shaders.adoc

ifndef::chapters[:chapters:]
ifndef::images[:images: images/]

// the [] in the URL messes up asciidoc
:max-compute-workgroup-size-link: https://vulkan.gpuinfo.org/displaydevicelimit.php?name=maxComputeWorkGroupSize&#91;0&#93;&platform=all
:max-compute-workgroup-count-link: https://vulkan.gpuinfo.org/displaydevicelimit.php?name=maxComputeWorkGroupCount&#91;0&#93;&platform=all

[[compute-shaders]]
= Compute Shaders

This chapter is **not** a "how to use compute shader" article, there are plenty of resources online around GPGPU and compute.

What this chapter is for is all the "Vulkan-ism", terms, etc that are associated with compute shaders.

There is also a xref:{chapters}decoder_ring.adoc[Decoder Ring] created to help people transition from other APIs that use different terminology.

[NOTE]
====
If you want to play around with a simple compute example, suggest taking a look at the link:https://github.com/charles-lunarg/vk-bootstrap/blob/main/example/simple_compute.cpp[vk-bootstrap sample].
====

== Coming from Vulkan Graphics

For those who are more familiar with graphics in Vulkan, compute will be a simple transition. Basically everything is the same except:

> - No vertex buffers, render passes, swapchains needed

Reviewer: While technically correct that they aren't required to execute the pipeline, it's worth mentioning that compute shaders often interact with these resources (e.g., writing data into a vertex buffer or post-processing a swapchain image as a storage image). This clarifies that they aren't "isolated" from the rest of the graphics pipeline.

> - Call `vkCmdDispatch` instead of `vkCmdDraw`

Reviewer: I'd recommend mentioning vkCmdDispatchIndirect here as well. Indirect dispatch is a cornerstone of advanced compute usage and GPU-driven pipelines, and it fits perfectly in a list of "Vulkan-isms."

- Use `vkCreateComputePipelines` instead of `vkCreateGraphicsPipelines`
- Make sure your `VkQueue` xref:{chapters}queues.adoc[supports] `VK_QUEUE_COMPUTE_BIT`

> - When binding descriptors or pipeline to your command buffer, make sure to use `VK_PIPELINE_BIND_POINT_COMPUTE`

Reviewer 2: ```suggestion
- When binding descriptors and pipelines to your command buffer, make sure to use `VK_PIPELINE_BIND_POINT_COMPUTE`
```

== SPIR-V Terminology

The smallest unit of work that is done is called an `invocation`. It is a "thread" or "lane" of work.

> `Invocations` are partitioned into `subgroups`, where `invocations` within a `subgroup` can synchronize and share data with each other efficiently. (See more in the xref:{chapters}subgroups.adoc[subgroup chapter])

Reviewer 2: "`Invocations` are partitioned into `subgroups`" almost implies that individual invocations are being cut up into subgroups. 

```suggestion
Multiple `Invocations` are organized into `subgroups`, where `invocations` within a `subgroup` can synchronize and share data with each other efficiently. (See more in the xref:{chapters}subgroups.adoc[subgroup chapter])
```

Next we have `workgroups` which is the smallest unit of work that an application can define. A `workgroup` is a collection of `invocations` that execute the same shader.

[NOTE]
====
While slightly annoying, Vulkan spec uses `WorkGroup` while the SPIR-V spec spells it as `Workgroup`. It has no significant meaning, other than a potential typo when going between the two.
====

=== Workgroup Size

Setting the `workgroup` size can be done in 3 ways

> 1. Using the `WorkgroupSize` built-in (link:https://godbolt.org/z/ees83eT7x[example])

Reviewer 3: Most developers wouldn't care about this detail and it would be more useful to show how it's declared in GLSL/HLSL/Slang.

Author: sure, part of this was the "pure" vulkan/spirv point-of-view, but can easily throw a "how to do it in GLSL/HLSL/Slang" example with it

2. Using the `LocalSize` execution mode (link:https://godbolt.org/z/3zn1Preb8[example])
3. Using the `LocalSizeId` execution mode (link:https://godbolt.org/z/dP7daqTas[example])

A few important things to note:

- The `WorkgroupSize` decoration will take precedence over any `LocalSize` or `LocalSizeId` in the same module.

> - `LocalSizeId` was added in `VK_KHR_maintenance4` (Vulkan 1.3) to allow the ability to use specialization constants to set the size.

Reviewer 2: ```suggestion
- `LocalSizeId` was added in the extension `VK_KHR_maintenance4` (made core in Vulkan 1.3) to allow the ability to use specialization constants to set the size.
```

> - There is a `maxComputeWorkGroupSize` limit how large the `X`, `Y`, and `Z` size can each be (most implementations {max-compute-workgroup-size-link}[support around 1024 for each dimension])

Reviewer 2: This almost implies that there is one limit that governs the size of all dimensions (separately) when in reality each dimension is a separate limit. Not sure how to make that clear or if its even worth the ink to clarify, but I mention it cause it caused me concern at first.

> - There is a `maxComputeWorkGroupInvocations` limit how large the product of `X` * `Y` * `Z` can be (most implementations link:https://vulkan.gpuinfo.org/displaydevicelimit.php?name=maxComputeWorkGroupInvocations&platform=all[support around 1024])

Reviewer 2: ```suggestion
- There is a `maxComputeWorkGroupSize` limit how large the `X`, `Y`, and `Z` size can each be. Most implementations {max-compute-workgroup-size-link}[support around 1024 for each dimension]
- There is a `maxComputeWorkGroupInvocations` limit how large the product of `X` * `Y` * `Z` can be. Most implementations link:https://vulkan.gpuinfo.org/displaydevicelimit.php?name=maxComputeWorkGroupInvocations&platform=all[support around 1024]
```

=== Local and Global Workgroups

> When a `vkCmdDispatch` is called, it sets the number of workgroups to dispatch. This produces a `global workgroup` space that the GPU will work on. Each single workgroup is a `local workgroup`. An `invocation` within a `local workgroup` can share data with other members of the `local workgroup` through shared variables and issue memory and control flow barriers to synchronize with other members of the `local workgroup`.

Reviewer 2: ```suggestion
When `vkCmdDispatch` is called, it sets the number of workgroups to dispatch. This produces a `global workgroup` space that the GPU will work on. Each single workgroup is a `local workgroup`. An `invocation` within a `local workgroup` can share data with other members of the `local workgroup` through shared variables as well as issue memory and control flow barriers to synchronize with other members of the `local workgroup`.
```
(too many ands in a row)

[NOTE]
====
There is a `maxComputeWorkGroupCount` limit link:{max-compute-workgroup-count-link}[some hardware] supports only 64k, but newer hardware can basically be unlimited here.
====

== Shared memory

When inside a single `local workgroup` "shared memory" can be used. In SPIR-V this is referenced with the `Workgroup` storage class.

Shared memory is essentially the "L1 cache you can control" in your compute shader and an important part of any performant shader.

There is a `maxComputeSharedMemorySize` limit (link:https://vulkan.gpuinfo.org/displaydevicelimit.php?name=maxComputeSharedMemorySize&platform=all[mainly around 32k bytes]) that needs to be accounted for.

=== Shared Memory Race Conditions

> It is very easy to get into race conditions with shared memory.

Reviewer 2: ```suggestion
It is very easy to have race conditions when using shared memory.
```

> The classic example is having multiple invocations initializing something to the same value.

Reviewer 2: ```suggestion
The classic example is when multiple invocations initialize something to the same value.
```

[source,glsl]
----
shared uint my_var;
void main() {

>     // All the invocations in the workgroup are going to try to write to the same memory.

Reviewer 3: I think it's confusing to lead with this example, and would be better to start with examples of cross-thread communication and using barrier() first.

Author: can you provide some GLSL snippets here (in the comment), I know you are more of an expert in this area and probably better at providing some better examples to provide

Reviewer 3: You could use a reduction as a simple example of cross-thread communication:

```
    shared float temp[BLOCK_SIZE];

    temp[tid] = x;
    [[unroll]] for (uint s = BLOCK_SIZE / 2; s > 0; s >>= 1) {
        barrier();
        float other = temp[tid ^ s];
        barrier();
        temp[tid] += other;
    }
```

----

```
shared float my_var[BLOCK_SIZE];

void reduction(float x) {
    uint tid = gl_GlobalInvocationID;

    my_var[tid] = x;
    for (uint s = BLOCK_SIZE / 2; s > 0; s >>= 1) {
        barrier();
        float other = my_var[tid ^ s];
        barrier();
        my_var[tid] += other;
    }
}
```

Author: sorry, we can add this in a follow up, this example makes is confusing... was is suppose to be an example of having a race condition?

    // RACE CONDITION
    my_var = 0;
}
----

If you are asking "why?", the "technically correct" answer is "because the link:https://docs.vulkan.org/spec/latest/appendices/memorymodel.html[memory model] says so".

When you do a weak store to a memory location, that invocation "owns" that memory location until synchronization occurs. The compiler **can** use that information and choose to reuse that location as temporary storage for another value.

> Luckily the "fix" is simple, make sure to use atomics

Reviewer 2: Why is 'fix' in quotes, is it not actually a fix? Or more of a 'this may be slower than you anticipate'?

[source,glsl]
----
shared uint my_var;
void main() {

>     atomicStore(temp, 0u, gl_ScopeWorkgroup, 0, 0);

Reviewer: There is a typo here—temp should be my_var to match the variable declared above.

}
----

Another option is to use a `OpControlBarrier` with `Workgroup` scope (link:https://godbolt.org/z/WcsvjYfPx[see online]).

[source,glsl]
----
layout(local_size_x = 32) in; // 32x1x1 workgroup
shared uint my_var[32]; // one slot for each invocation

void main() {
    my_var[gl_LocalInvocationIndex] = 0;
    barrier(); // will generate an OpControlBarrier for you
    uint x = my_var[gl_LocalInvocationIndex ^ 1];
}
----

==== Detecting shared memory data races

Luckily this problem can be caught automatically using the link:https://github.com/KhronosGroup/Vulkan-ValidationLayers/blob/main/docs/gpu_validation.md[GPU-AV] feature in Vulkan Validation Layers!

> As of March 2026 (TODO - Add SDK version when released in May) when using GPU-AV, it will attempt to detect these races for you. There are a link:https://github.com/KhronosGroup/Vulkan-ValidationLayers/blob/main/layers/gpuav/shaders/instrumentation/shared_memory_data_race.comp#L47[few limitations], but highly suggest trying out if having strange issues around your shared memory accesses.

Reviewer 2: ```suggestion
As of March 2026 (TODO - Add SDK version when released in May), GPU-AV will attempt to detect these races for you. There are a link:https://github.com/KhronosGroup/Vulkan-ValidationLayers/blob/main/layers/gpuav/shaders/instrumentation/shared_memory_data_race.comp#L47[few limitations], but highly suggest trying out if having strange issues around your shared memory accesses.
```
